File: src/main/java/chapter7/RetryExample.java
package chapter7;

import common.CommonUtils;
import common.Log;
import common.OkHttpHelper;
import io.reactivex.Observable;
import io.reactivex.ObservableEmitter;
import io.reactivex.schedulers.Schedulers;

import java.util.concurrent.TimeUnit;


public class RetryExample {

    public void retry5() {
        CommonUtils.exampleStart();

        String url = "https://api.github.com/zen";
        Observable<String> source = Observable.just(url)
                .map(OkHttpHelper::get)
                .retry(5)
                .onErrorReturn(e -> CommonUtils.ERROR_CODE);

        source.subscribe(data -> Log.it("result : " + data));
    }

    public void retryWithDelay() {
        final int RETRY_MAX = 5;
        final int RETRY_DELAY = 1000;

        CommonUtils.exampleStart();

        String url = " https://api.github.com/zen";
        Observable<String> source = Observable.just(url)
                .map(OkHttpHelper::get)
                .retry((retryCnt, e) -> {
                    Log.it("retryCnt = " + retryCnt);
                    CommonUtils.sleep(RETRY_DELAY);

                    return retryCnt < RETRY_MAX;
                })
                .onErrorReturn(e -> CommonUtils.ERROR_CODE);

        source.subscribe(data -> Log.it("result : " + data));
    }

    public void retryUntil() {
        CommonUtils.exampleStart();

        String url = "https://api.github.com/zen";
        Observable<String> source = Observable.just(url)
                .map(OkHttpHelper::get)
                .subscribeOn(Schedulers.io())
                .retryUntil(() -> {
                    if(CommonUtils.isNetworkAvailable()) {
                        return true;
                    }
                    CommonUtils.sleep(1000);
                    return false;
                });
        source.subscribe(Log::it);

        CommonUtils.sleep(5000);
    }

    public void retryWhen() {
        Observable.create((ObservableEmitter<String> emitter) -> {
            emitter.onError(new RuntimeException("always fails"));
        }).retryWhen(attempts -> {
            return attempts.zipWith(Observable.range(1, 3), (n, i) -> i)
                    .flatMap(i -> {
                        Log.it("delay retry by " + i + " seconds");
                        return Observable.timer(1 , TimeUnit.SECONDS);
                    });
        }).blockingForEach(Log::it);
    }

    public static void main(String[] args) {
        RetryExample demo = new RetryExample();
        // demo.retry5();
        // demo.retryWithDelay();
        // demo.retryUntil();
        demo.retryWhen();
    }

}

File: src/main/kotlin/chapter7/RetryExample_.kt
package chapter7

import common.CommonUtils_
import common.Log_
import common.OkHttpHelper_
import io.reactivex.Observable
import io.reactivex.functions.BiFunction
import io.reactivex.schedulers.Schedulers
import java.util.concurrent.TimeUnit

class RetryExample_ {

    fun retry5() {
        CommonUtils_.exampleStart()

        val url = "https://api.github.com/zen"
        val source = Observable.just(url)
                .map { OkHttpHelper_.get(it) }
                .retry(5)
                .onErrorReturn { CommonUtils_.ERROR_CODE }

        source.subscribe { Log_.it("result : $it") }
    }

    fun retryWithDelay() {
        val RETRY_MAX = 5
        val RETRY_DELAY = 1000

        CommonUtils_.exampleStart()

        val url = " https://api.github.com/zen"
        val source = Observable.just(url)
                .map { OkHttpHelper_.get(it) }
                .retry { retryCnt, _ ->
                    Log_.it("retryCnt = $retryCnt")
                    CommonUtils_.sleep(RETRY_DELAY)
                    retryCnt < RETRY_MAX
                }
                .onErrorReturn { CommonUtils_.ERROR_CODE }

        source.subscribe { Log_.it("result : $it") }
    }

    fun retryUntil() {
        CommonUtils_.exampleStart()

        val url = "https://api.github.com/zen"
        val source = Observable.just(url)
                .map { OkHttpHelper_.get(it) }
                .subscribeOn(Schedulers.io())
                .retryUntil {
                    if (CommonUtils_.isNetworkAvailable()) {
                        true
                    }
                    CommonUtils_.sleep(1000)
                    false
                }
        source.subscribe { Log_.it(it) }

        CommonUtils_.sleep(5000)
    }

    fun retryWhen() {
        Observable.create<String> {
            it.onError(RuntimeException("always fails"))
        }.retryWhen { attempts ->
            attempts.zipWith(Observable.range(1, 3), BiFunction { _: Throwable, i: Int -> i })
                    .flatMap({ i ->
                        Log_.it("delay retry by $i seconds")
                        Observable.timer(1, TimeUnit.SECONDS)
                    })
        }.blockingForEach { Log_.it(it) }
    }

}


fun main(args: Array<String>) {
    val demo = RetryExample_()
    // demo.retry5()
    // demo.retryWithDelay()
    // demo.retryUntil()
    demo.retryWhen()
}
